### PR TITLE
Closes #5815 Correctly clear cache for URLs whose permalink contains uppercase letters in non-latin alphabets

### LINKED ISSUE #5815: Cache isn't cleared for URLs whose permalink contains uppercase letters in non-latin alphabets
**Before submitting an issue please check that you’ve completed the following steps:**
- [x] Made sure you’re on the latest version
- [x]  Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**

URLs that should have their cache cleared after updating a post aren't cleared if they contain an uppercase character in a non-latin alphabet.

**To Reproduce**
1. Install a plugin that allows you to have custom permalinks [(Permalink Manager Lite)](https://wordpress.org/plugins/permalink-manager/)
2. Create and publish a post and change its URL to contain an uppercase non-latin character (Γυναικείο, Производ)
Image here: https://jmp.sh/J06a6ta2
3. Load the post once to create a cache.
4. Change the post and update it.
5. Post page doesn't reflect the update as the cache files aren't cleared.

**Expected behavior**
When a user clicks the update button, the post's cache file should be deleted so it reflects the changes.

**Additional context**

The issue is happening when we store the URL on the server, we do it in lowercase [here](https://github.com/wp-media/wp-rocket/blob/60b76bc8bc9a366bf5b9644894df8203bba103fd/inc/classes/Buffer/class-cache.php#L392), but then we search for the URL with the uppercase characters [here](https://github.com/wp-media/wp-rocket/blob/0015d58087695811bf42138eba9dd914741a9519/inc/functions/files.php#L535) because we assume that the URL will be all lowercase.

Adding `$entry = strtolower( $entry );` after `$entry = strtolower( $entry );` [here](https://github.com/wp-media/wp-rocket/blob/0015d58087695811bf42138eba9dd914741a9519/inc/functions/files.php#L535) fixes the issue.

Related ticket: https://secure.helpscout.net/conversation/2177113016/406819?folderId=7469349
Initial Slack discussion: https://wp-media.slack.com/archives/C43T1AYMQ/p1678448004086059
Possible related issue #5561 

Currently working on a workaround for the user.

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [x] Scope a solution
- [x] Estimate the effort

### PR DESCRIPTION
**Pull request by jeawhanlee**
## Description
Transforms url character to lower case before deleting cache.

Fixes #5815 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

No

## How Has This Been Tested?

Manual Test

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings

### CODE REVIEW

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Thanks for the PR. Currently , on branch (after merging dev to the branch locally)
, can see the cache isnot cleared to the post in the following case (can you please check):
1- Disable preload 
2- Set permalinks to `/%postname%/роизвод`
3- cache a post
4- edit post and check cache => the cache of the post isnot cleared

**Review by mostafa-hisham — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee While exploratory test can see that we have a regression in the following scenario (was working fine on 3.13.0.1)
1- post URL with slug `γυναικείο` 
2- permalinks `/%postname%`
3- Preload is disabled
4- cache the post url
5- edit the post and check cache => cache isnot cleared
can you please check that?

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Thanks for the update. the slug case is working fine now. However, if permalinks contains capital Latin letter, cache won't be cleared when edit the post. can you please check that?
steps:
1- while preload is disabled, set permalinks `/%postname%/Test/`
2- cache a post i.e `https://newer.rocketlabsqa.ovh/new-test/Test/`
3- edit post and check cache
Expected:
Cache is cleared as same as with Capital non-Latin letters

File: inc/functions/files.php [hunk 1]
```diff
@@ -596,7 +595,17 @@ function rocket_clean_files( $urls, $filesystem = null ) {
 
 		if ( ! empty( $parsed_url['host'] ) ) {
 			foreach ( _rocket_get_cache_dirs( $parsed_url['host'], $cache_path ) as $dir ) {
+				// Encode Non-latin characters if found in url path.
+				if ( false !== preg_match_all( '/(?<non_latin>[^\x00-\x7F]+)/', $parsed_url['path'], $matches ) ) {
+					$encode_non_latin = function( $non_latin ) {
```
Comment (remyperona) [inline inc/functions/files.php]:
```suggestion
					$cb_encode_non_latin = function( $non_latin ) {
```

File: inc/functions/files.php [hunk 2]
```diff
@@ -596,7 +595,17 @@ function rocket_clean_files( $urls, $filesystem = null ) {
...
+						return strtolower( rawurlencode( $non_latin ) );
+					};
+
+					$parsed_url['path'] = str_replace( $matches['non_latin'], array_map( $encode_non_latin, $matches['non_latin'] ), $parsed_url['path'] );
```
Comment (remyperona) [inline inc/functions/files.php]:
```suggestion
					$parsed_url['path'] = str_replace( $matches['non_latin'], array_map( $cb_encode_non_latin, $matches['non_latin'] ), $parsed_url['path'] );
```

**Review by remyperona — CHANGES_REQUESTED** [review]
small naming change to make it more clear the variable is a callback function

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee The capital Latin char issue is fixed now. However, if permalinks contains non-latin char then we cache post then click purge this URL from post view, the post cache won't be cleared (as same as on trunk). can you please check?

**Comment by piotrbak** [comment]
@jeawhanlee Could you check the effort of what @Mai-Saad mentioned? Do you feel it can introduce any bigger risk?

**Comment by jeawhanlee** [comment]
> @jeawhanlee Could you check the effort of what @Mai-Saad mentioned? Do you feel it can introduce any bigger risk?

I'm checking this...

**Comment by jeawhanlee** [comment]
@piotrbak @Mai-Saad To fix this we have to revert the previous commit on this [PR](https://github.com/wp-media/wp-rocket/pull/5852/commits/340995d67195bab4e996cdc392a39e9ddeca246f) that tries to fix @Mai-Saad comment [here](https://github.com/wp-media/wp-rocket/pull/5852#issuecomment-1513104271) as I feel that `purge this url` from post view is something that will be used more often.
what do you think?

**Comment by piotrbak** [comment]
@jeawhanlee The comment you mentioned is about the Latin char while the question [here](https://github.com/wp-media/wp-rocket/pull/5852#issuecomment-1543724893) is about the non-Latin one, right?

We should be able to clear the cache for the following scenarios:
1) non-Latin uppercase character present in the permalink or post title
2) non-Latin permalink or post title contains only lowercase characters
3) Latin uppercase character present in the permalink or post title
4) Latin permalink or post title contains only lowercase characters

Would be great if all those scenarios worked with the trailing slash at the end or without it in the permalinks settings + when `updating the post` or using `Purge this URL`. It's important not to introduce any regression in this PR though, as we're fixing a problem that's not touching majority of our customers.

We can brainstorm the problem here to find a solution that would cover everything.

**Comment by piotrbak** [comment]
@vmanthos Agreed!

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Can you please check these cases?
Case1:
1- Preload is off then Permalinks set to /%postname%/дÆÈÜ
2- We visit post URL [https://new.rocketlabsqa.ovh/test-post2/дÆÈÜ](https://new.rocketlabsqa.ovh/test-post2/%D0%B4%C3%86%C3%88%C3%9C%60) incognito => the following entry added to cache table as completed https://new.rocketlabsqa.ovh/test-post2/%D0%B4%C3%86%C3%88%C3%9C
3- edit the post => another entry [https://new.rocketlabsqa.ovh/test-post2/дÆÈÜ](https://new.rocketlabsqa.ovh/test-post2/%D0%B4%C3%86%C3%88%C3%9C%60) added to the database that will remain pending, and if we visit the post URL, the link in step 2 will update the last accessed

Note: same on the trunk, also on trunk purge this URL won't work for the link in step 2, however, it's working on the branch.

Case2:
1- Preload is disabled and the permalink is set to a mix of Latin/non-latin char i.e `Testäёåñô`
2- Cache a post `https://new.rocketlabsqa.ovh/test-post/Testäёåñô`
3- purge this URL or clear cache from the posts list or edit the post => cache won't be cleared (note, if we purge this URL, the status will change in the table but the cache folder is not cleared while clear cache from posts list /edit post will behave like case 1 but without clearing cache)



@piotrbak what do you think?

**Update: As per discussion with @piotrbak , for case 1, Since we already fixed the main issue of cache clear and the problem is already on trunk regarding cache table then we will open new ticket for this**

**Comment by piotrbak** [comment]
Thanks @Mai-Saad, @jeawhanlee could you check the second case? How complex is this one?

**Review by CrochetFeve0251 — APPROVED** [review]

**Comment by Mai-Saad** [comment]
@jeawhanlee Thanks for the update, now the cache is cleared with permalinks containing [Latin/non-Latin](https://docs.google.com/spreadsheets/d/1mwC9zaEASEyLrodEDeAaCdE57_D81F58Mn7MYpoyn6s/edit#gid=0) char when edit the post, clear this cache, purge this URL and clear the whole cache.
However, When we enable RUCSS, once used CSS is created, we aren't clearing the cache folder of related post although the cache table status changed to pending (while preload is off) while permalinks contain non-latin char i.e `/%postname%/ПÇŞĞIİÖÜ` @piotrbak what do you think?

**Comment by piotrbak** [comment]
@Mai-Saad Is it only about the `cache folder` or the cache files itselves?

**Comment by Mai-Saad** [comment]
@piotrbak the cache files.
1- Disable preload
2- Use permalinks` /%postname%/ПÇŞĞIİÖÜ`
3- Cache a post i.e https://new.rocketlabsqa.ovh/test-post-2/ПÇŞĞIİÖÜ/
4- Enable RUCSS 
5- Once Used CSS is completed at the used CSS table => check the cache table and cache folder, files are still in the cache folder while the status switched to pending at the cache table

**Comment by piotrbak** [comment]
@Mai-Saad That sounds like an edge case, I think we're okay here.

**Comment by Mai-Saad** [comment]
ok, another note:
- when permalinks contain non-latin char, then we switch post to draft => cache is cleared but status at cache table remains completed
- For case 1, is occurring too when trash a post or switch a post to draft
- Already on the trunk (while preload off), when we delete the post, we are not removing it from the cache table

**Review by Mai-Saad — APPROVED** [review]
As agreed, mentioned points will remain as is. Working fine then

**Pull request merged**